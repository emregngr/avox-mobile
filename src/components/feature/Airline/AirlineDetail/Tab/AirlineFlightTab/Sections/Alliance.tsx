import { MaterialCommunityIcons } from '@expo/vector-icons'
import React from 'react'
import { View } from 'react-native'

import { ThemedText } from '@/components/common/ThemedText'
import { AirlineSectionRow } from '@/components/feature/Airline/AirlineDetail/AirlineSectionRow'

interface AllianceProps {
  alliance: string
  iconColor: string
  title: string
}

export const Alliance = ({ alliance, iconColor, title }: AllianceProps) => (
  <AirlineSectionRow title={title}>
    <View className="items-center my-5">
      <MaterialCommunityIcons color={iconColor} name="shield" size={40} />

      <ThemedText className="mt-2.5" color="text-90" type="h4">
        {alliance}
      </ThemedText>
    </View>
  </AirlineSectionRow>
)
