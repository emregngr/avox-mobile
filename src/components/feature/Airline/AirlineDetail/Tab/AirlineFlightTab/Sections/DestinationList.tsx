import React from 'react'
import { View } from 'react-native'

import { AirlineSectionRow } from '@/components/feature/Airline/AirlineDetail/AirlineSectionRow'
import { DestinationRowCard } from '@/components/feature/Airline/AirlineDetail/Tab/AirlineFlightTab/Cards/DestinationRowCard'

interface DestinationListProps {
  destinations: string[]
  iconColor: string
  title: string
}

export const DestinationList = ({ destinations, iconColor, title }: DestinationListProps) => (
  <AirlineSectionRow title={title}>
    <View className="flex-row flex-wrap -mx-1">
      {destinations?.map((destination, index) => (
        <DestinationRowCard destination={destination} iconColor={iconColor} key={index} />
      ))}
    </View>
  </AirlineSectionRow>
)
