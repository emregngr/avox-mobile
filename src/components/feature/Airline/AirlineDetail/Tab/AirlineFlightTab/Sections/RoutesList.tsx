import React from 'react'

import { AirlineSectionRow } from '@/components/feature/Airline/AirlineDetail/AirlineSectionRow'
import { RouteRowCard } from '@/components/feature/Airline/AirlineDetail/Tab/AirlineFlightTab/Cards/RouteRowCard'
import type { AirlineRoute } from '@/types/feature/airline'

interface RoutesListProps {
  iconColor: string
  routes: AirlineRoute[]
  title: string
}

export const RoutesList = ({ iconColor, routes, title }: RoutesListProps) => (
  <AirlineSectionRow title={title}>
    {routes?.map((route, index) => (
      <RouteRowCard
        destinationIata={route.destinationIata}
        iconColor={iconColor}
        key={index}
        origin={route.origin}
      />
    ))}
  </AirlineSectionRow>
)
