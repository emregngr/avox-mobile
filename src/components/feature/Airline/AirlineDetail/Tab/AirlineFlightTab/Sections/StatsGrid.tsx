import React from 'react'
import { View } from 'react-native'

import { StatsCard } from '@/components/feature/Airline/AirlineDetail/Tab/AirlineFlightTab/Cards/StatsCard'

interface StatsGridProps {
  iconColor: string
  localeStrings: {
    country: string
    domesticDestinations: string
    internationalDestination: string
    totalDestination: string
  }
  stats: {
    destinationCount: number
    destinationCountries: number
    domesticConnections: number
    internationalConnections: number
  }
}

export const StatsGrid = ({ iconColor, localeStrings, stats }: StatsGridProps) => (
  <>
    <View className="flex-row gap-x-4">
      <StatsCard
        iconColor={iconColor}
        iconName="flag"
        label={localeStrings.totalDestination}
        value={stats.destinationCount}
      />
      <StatsCard
        iconColor={iconColor}
        iconName="earth"
        label={localeStrings.country}
        value={stats.destinationCountries}
      />
    </View>

    <View className="flex-row gap-x-4">
      <StatsCard
        iconColor={iconColor}
        iconName="home"
        label={localeStrings.domesticDestinations}
        value={stats.domesticConnections}
      />
      <StatsCard
        iconColor={iconColor}
        iconName="globe"
        label={localeStrings.internationalDestination}
        value={stats.internationalConnections}
      />
    </View>
  </>
)
