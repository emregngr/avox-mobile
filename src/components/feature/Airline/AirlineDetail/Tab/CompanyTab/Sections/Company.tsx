import React, { useMemo } from 'react'

import { AirlineRowItem } from '@/components/feature/Airline/AirlineDetail/AirlineRowItem'
import { AirlineSectionRow } from '@/components/feature/Airline/AirlineDetail/AirlineSectionRow'
import { getLocale } from '@/locales/i18next'
import useLocaleStore from '@/store/locale'
import type { Airline } from '@/types/feature/airline'
import { formatNumber } from '@/utils/feature/formatNumber'

interface CompanyProps {
  airlineData: Airline
}

export const Company = ({ airlineData }: CompanyProps) => {
  const { selectedLocale } = useLocaleStore()

  const { companyInfo, isoCountry, isoRegion, operations } = airlineData || {}
  const { employeeCount, foundingYear, parentCompany, passengerCapacity } = companyInfo || {}
  const { businessModel, businessType, skytraxRating, slogan } = operations || {}

  const localeStrings = useMemo(
    () => ({
      airlineInfo: getLocale('airlineInfo'),
      businessModel: getLocale('businessModel'),
      businessType: getLocale('businessType'),
      isoCountryCode: getLocale('isoCountryCode'),
      isoRegionCode: getLocale('isoRegionCode'),
      millionPerYear: getLocale('million/year'),
      numberOfEmployees: getLocale('numberOfEmployees'),
      numberOfPassengers: getLocale('numberOfPassengers'),
      parentCompany: getLocale('parentCompany'),
      skytraxRating: getLocale('skytraxRating'),
      slogan: getLocale('slogan'),
      year: getLocale('year'),
      yearOfEstablishment: getLocale('yearOfEstablishment'),
    }),
    [selectedLocale],
  )

  return (
    <AirlineSectionRow title={localeStrings.airlineInfo}>
      <AirlineRowItem
        icon="calendar"
        label={localeStrings.yearOfEstablishment}
        value={`${foundingYear} (${new Date().getFullYear() - parseInt(foundingYear)} ${localeStrings.year})`}
      />

      <AirlineRowItem icon="business" label={localeStrings.parentCompany} value={parentCompany} />

      <AirlineRowItem
        icon="people"
        label={localeStrings.numberOfPassengers}
        value={`${passengerCapacity} ${localeStrings.millionPerYear}`}
      />

      <AirlineRowItem
        icon="people"
        label={localeStrings.numberOfEmployees}
        value={formatNumber(employeeCount)}
      />

      <AirlineRowItem
        className="capitalize"
        icon="briefcase"
        label={localeStrings.businessModel}
        value={businessModel?.replace('_', ' ')}
      />

      <AirlineRowItem
        className="capitalize"
        icon="airplane"
        label={localeStrings.businessType}
        value={businessType?.replace('_', ' ')}
      />

      <AirlineRowItem icon="flag" label={localeStrings.isoCountryCode} value={isoCountry} />

      <AirlineRowItem icon="location" label={localeStrings.isoRegionCode} value={isoRegion} />

      <AirlineRowItem
        icon="star"
        label={localeStrings.skytraxRating}
        value={`${skytraxRating?.toFixed(1)} / 5`}
      />

      <AirlineRowItem icon="chatbox-ellipses" label={localeStrings.slogan} value={slogan} />
    </AirlineSectionRow>
  )
}
