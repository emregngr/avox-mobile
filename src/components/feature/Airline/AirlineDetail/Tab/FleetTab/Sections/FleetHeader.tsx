import { MaterialCommunityIcons } from '@expo/vector-icons'
import React, { useMemo } from 'react'
import { View } from 'react-native'

import { ThemedText } from '@/components/common/ThemedText'
import useThemeStore from '@/store/theme'
import { themeColors } from '@/themes'

interface FleetHeaderProps {
  fleetDetailText: string
}

export const FleetHeader = ({ fleetDetailText }: FleetHeaderProps) => {
  const { selectedTheme } = useThemeStore()

  const colors = useMemo(() => themeColors?.[selectedTheme], [selectedTheme])

  return (
    <View className="flex-row items-center">
      <MaterialCommunityIcons color={colors?.onPrimary100} name="airplane" size={20} />

      <ThemedText className="ml-3" color="text-100" type="h3">
        {fleetDetailText}
      </ThemedText>
    </View>
  )
}
