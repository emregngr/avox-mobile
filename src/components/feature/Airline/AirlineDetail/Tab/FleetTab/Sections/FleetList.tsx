import React, { useCallback } from 'react'
import type { ListRenderItem } from 'react-native'
import { FlatList } from 'react-native'

import { AirplaneRowCard } from '@/components/feature/Airline/AirlineDetail/Tab/FleetTab/Card/AirplaneRowCard'
import type { ThemeColors } from '@/themes'
import type { Airplane } from '@/types/feature/airline'

const ITEMS_PER_PAGE = 3
const itemHeight = 200

interface FleetListProps {
  airplanes: Airplane[]
  colors: ThemeColors
  onImagePress: (airplaneType: string, imageKey: string) => void
  region: string
  totalAirplane: number
}

export const FleetList = ({
  airplanes,
  colors,
  onImagePress,
  region,
  totalAirplane,
}: FleetListProps) => {
  const renderAirplaneItem = useCallback<ListRenderItem<Airplane>>(
    ({ item }) => (
      <AirplaneRowCard
        airplane={item}
        colors={colors}
        onImagePress={onImagePress}
        region={region}
        totalAirplane={totalAirplane}
      />
    ),
    [totalAirplane, region, colors, onImagePress],
  )

  const keyExtractor = useCallback((item: Airplane) => item?.type, [])

  const getItemLayout = useCallback(
    (data: any, index: number) => ({
      index,
      length: itemHeight,
      offset: itemHeight * index,
    }),
    [],
  )

  return (
    <FlatList
      data={airplanes}
      getItemLayout={getItemLayout}
      initialNumToRender={ITEMS_PER_PAGE}
      keyExtractor={keyExtractor}
      maxToRenderPerBatch={ITEMS_PER_PAGE}
      renderItem={renderAirplaneItem}
      scrollEventThrottle={16}
      showsVerticalScrollIndicator={false}
      updateCellsBatchingPeriod={200}
      windowSize={5}
      removeClippedSubviews
    />
  )
}
