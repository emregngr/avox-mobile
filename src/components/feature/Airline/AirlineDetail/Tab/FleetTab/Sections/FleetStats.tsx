import React, { memo, useMemo } from 'react'
import { View } from 'react-native'

import { ThemedText } from '@/components/common/ThemedText'
import { getLocale } from '@/locales/i18next'
import useLocaleStore from '@/store/locale'

interface FleetStatsProps {
  airplaneTypeCount: number
  averageAgeYears: number
  totalAirplane: number
}

export const FleetStats = memo(
  ({ airplaneTypeCount, averageAgeYears, totalAirplane }: FleetStatsProps) => {
    const { selectedLocale } = useLocaleStore()

    const localeStrings = useMemo(
      () => ({
        airplaneType: getLocale('airplaneType'),
        averageAgeYears: getLocale('averageAgeYears'),
        totalAirplane: getLocale('totalAirplane'),
      }),
      [selectedLocale],
    )

    return (
      <View className="flex-row p-4 rounded-xl overflow-hidden bg-background-secondary">
        <View className="flex-1 items-center">
          <ThemedText className="mb-1" color="text-100" type="h1">
            {totalAirplane}
          </ThemedText>
          <ThemedText
            color="text-90" lineBreakMode="tail" numberOfLines={2}
            type="body3" center
          >
            {localeStrings.totalAirplane}
          </ThemedText>
        </View>

        <View className="w-px h-12 bg-background-tertiary" />

        <View className="flex-1 items-center">
          <ThemedText className="mb-1" color="text-100" type="h1">
            {averageAgeYears}
          </ThemedText>
          <ThemedText
            color="text-90" lineBreakMode="tail" numberOfLines={2}
            type="body3" center
          >
            {localeStrings.averageAgeYears}
          </ThemedText>
        </View>

        <View className="w-px h-12 bg-background-tertiary" />

        <View className="flex-1 items-center">
          <ThemedText className="mb-1" color="text-100" type="h1">
            {airplaneTypeCount}
          </ThemedText>
          <ThemedText
            color="text-90" lineBreakMode="tail" numberOfLines={2}
            type="body3" center
          >
            {localeStrings.airplaneType}
          </ThemedText>
        </View>
      </View>
    )
  },
)
