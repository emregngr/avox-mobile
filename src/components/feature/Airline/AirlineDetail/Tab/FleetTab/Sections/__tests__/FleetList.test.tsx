import { fireEvent, render } from '@testing-library/react-native'
import React from 'react'

import { FleetList } from '@/components/feature/Airline/AirlineDetail/Tab/FleetTab/Sections/FleetList'

const mockAirplaneRowCardSpy = jest.fn()

jest.mock('@/components/feature/Airline/AirlineDetail/Tab/FleetTab/Card/AirplaneRowCard', () => {
  const { Pressable, View, Text } = require('react-native')
  return {
    AirplaneRowCard: (props: { airplane: any; onImagePress: any }) => {
      mockAirplaneRowCardSpy(props)
      const { airplane, onImagePress } = props
      return (
        <View testID={`airplane-row-${airplane.type}`}>
          <Pressable
            onPress={() => onImagePress(airplane.type, 'some-image-key')}
            testID={`airplane-image-${airplane.type}`}
          >
            <Text>{airplane.type}</Text>
          </Pressable>
        </View>
      )
    },
  }
})

const mockedAirplanes: any = [
  { id: '1', type: 'B737-800', registration: 'TC-JHM' },
  { id: '2', type: 'A320-200', registration: 'TC-JPT' },
  { id: '3', type: 'B777-300ER', registration: 'TC-LJA' },
]

const mockedOnImagePress = jest.fn()

const mockedDefaultProps: any = {
  airplanes: mockedAirplanes,
  onImagePress: mockedOnImagePress,
  region: 'EU',
  totalAirplane: mockedAirplanes.length,
}

describe('FleetList Component', () => {
  it('should render the correct number of airplane cards', () => {
    const { getAllByTestId } = render(<FleetList {...mockedDefaultProps} />)
    const airplaneCards = getAllByTestId(/airplane-row-/)
    expect(airplaneCards).toHaveLength(mockedAirplanes.length)
  })

  it('should render nothing when the airplanes array is empty', () => {
    const { queryAllByTestId } = render(<FleetList {...mockedDefaultProps} airplanes={[]} />)
    const airplaneCards = queryAllByTestId(/airplane-row-/)
    expect(airplaneCards).toHaveLength(0)
  })

  it('should pass the correct props down to each AirplaneRowCard', () => {
    render(<FleetList {...mockedDefaultProps} />)
    expect(mockAirplaneRowCardSpy).toHaveBeenCalledWith(
      expect.objectContaining({
        airplane: mockedAirplanes[0],
        region: 'EU',
        totalAirplane: mockedAirplanes.length,
      }),
    )
  })

  it('should call onImagePress with correct arguments when a card is pressed', () => {
    const { getByTestId } = render(<FleetList {...mockedDefaultProps} />)
    const firstAirplaneImage = getByTestId(`airplane-image-${mockedAirplanes[0].type}`)
    fireEvent.press(firstAirplaneImage)
    expect(mockedOnImagePress).toHaveBeenCalledTimes(1)
    expect(mockedOnImagePress).toHaveBeenCalledWith(mockedAirplanes[0].type, 'some-image-key')
  })

  it('should have correct performance props set on FlatList', () => {
    const { getByTestId } = render(<FleetList {...mockedDefaultProps} />)
    const flatList = getByTestId('fleet-list')
    expect(flatList.props.initialNumToRender).toBe(4)
    expect(flatList.props.maxToRenderPerBatch).toBe(3)
    expect(flatList.props.windowSize).toBe(7)
    expect(flatList.props.updateCellsBatchingPeriod).toBe(150)
    expect(flatList.props.removeClippedSubviews).toBe(true)
  })
})

describe('FleetList Component Snapshot', () => {
  it('should render the FleetList Component successfully', () => {
    const { toJSON } = render(<FleetList {...mockedDefaultProps} />)

    expect(toJSON()).toMatchSnapshot()
  })
})
