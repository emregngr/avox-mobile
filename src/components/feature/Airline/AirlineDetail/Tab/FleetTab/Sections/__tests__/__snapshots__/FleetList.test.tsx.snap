// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`FleetList Component Snapshot should render the FleetList Component successfully 1`] = `
<RCTScrollView
  data={
    [
      {
        "id": "1",
        "registration": "TC-JHM",
        "type": "B737-800",
      },
      {
        "id": "2",
        "registration": "TC-JPT",
        "type": "A320-200",
      },
      {
        "id": "3",
        "registration": "TC-LJA",
        "type": "B777-300ER",
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  getItemLayout={[Function]}
  initialNumToRender={4}
  keyExtractor={[Function]}
  maxToRenderPerBatch={3}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={true}
  renderItem={[Function]}
  scrollEventThrottle={16}
  showsVerticalScrollIndicator={false}
  stickyHeaderIndices={[]}
  testID="fleet-list"
  updateCellsBatchingPeriod={150}
  viewabilityConfigCallbackPairs={[]}
  windowSize={7}
>
  <View>
    <View
      onFocusCapture={[Function]}
      style={null}
    >
      <View
        testID="airplane-row-B737-800"
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          testID="airplane-image-B737-800"
        >
          <Text>
            B737-800
          </Text>
        </View>
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      style={null}
    >
      <View
        testID="airplane-row-A320-200"
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          testID="airplane-image-A320-200"
        >
          <Text>
            A320-200
          </Text>
        </View>
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      style={null}
    >
      <View
        testID="airplane-row-B777-300ER"
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          testID="airplane-image-B777-300ER"
        >
          <Text>
            B777-300ER
          </Text>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
